Bottom: 090a807e09a659a09d49a7c5105ec0c95eaddffd
Top:    b8687aaba11f225806e46e6ff6a5811118a2ffa9
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2011-04-18 23:14:54 +0100

Don't abuse CrosshairOn() API to force a redraw.

Expose DrawAttached() to the HIDs so they don't have to cheat.

Make DrawAttached() and DrawMark() a NOP if the relevant item is not
being shown. Ie. if !Crosshair.On, both functions draw nothing. If
!Marked.status, DrawMark() draws nothing.

A minor change is required in CrosshairOff() to XOR un-draw before
switching the flag to off.


---

diff --git a/src/crosshair.c b/src/crosshair.c
index eb19257..7d9fcb4 100644
--- a/src/crosshair.c
+++ b/src/crosshair.c
@@ -565,10 +565,14 @@ XORDrawMoveOrCopyObject (void)
 /* ---------------------------------------------------------------------------
  * draws additional stuff that follows the crosshair
  */
-static void
+void
 DrawAttached (void)
 {
   BDimension s;
+
+  if (!Crosshair.On)
+    return;
+
   switch (Settings.Mode)
     {
     case VIA_MODE:
@@ -684,6 +688,28 @@ DrawAttached (void)
     }
 }
 
+
+/* --------------------------------------------------------------------------
+ * draw the marker position
+ */
+void
+DrawMark (void)
+{
+  /* Mark is not drawn when the crosshair is off, or when it is not set */
+  if (!Crosshair.On || !Marked.status)
+    return;
+
+  gui->draw_line (Crosshair.GC,
+                  Marked.X - MARK_SIZE,
+                  Marked.Y - MARK_SIZE,
+                  Marked.X + MARK_SIZE, Marked.Y + MARK_SIZE);
+  gui->draw_line (Crosshair.GC,
+                  Marked.X + MARK_SIZE,
+                  Marked.Y - MARK_SIZE,
+                  Marked.X - MARK_SIZE, Marked.Y + MARK_SIZE);
+}
+
+
 /* ---------------------------------------------------------------------------
  * switches crosshair on
  */
@@ -706,9 +732,9 @@ CrosshairOff (void)
 {
   if (Crosshair.On)
     {
-      Crosshair.On = false;
       DrawAttached ();
       DrawMark ();
+      Crosshair.On = false;
     }
 }
 
@@ -1086,26 +1112,6 @@ SetCrosshairRange (LocationType MinX, LocationType MinY, LocationType MaxX,
   MoveCrosshairRelative (0, 0);
 }
 
-/* --------------------------------------------------------------------------
- * draw the marker position
- * if argument is true, draw only if it is visible, otherwise draw it regardless
- */
-void
-DrawMark (void)
-{
-  if (Marked.status)
-    {
-      gui->draw_line (Crosshair.GC,
-		      Marked.X - MARK_SIZE,
-		      Marked.Y - MARK_SIZE,
-		      Marked.X + MARK_SIZE, Marked.Y + MARK_SIZE);
-      gui->draw_line (Crosshair.GC,
-		      Marked.X + MARK_SIZE,
-		      Marked.Y - MARK_SIZE,
-		      Marked.X - MARK_SIZE, Marked.Y + MARK_SIZE);
-    }
-}
-
 /* ---------------------------------------------------------------------------
  * initializes crosshair stuff
  * clears the struct, allocates to graphical contexts and
diff --git a/src/crosshair.h b/src/crosshair.h
index 4573e98..ff16b1e 100644
--- a/src/crosshair.h
+++ b/src/crosshair.h
@@ -46,18 +46,18 @@
 #define	STATE_SECOND	1
 #define	STATE_THIRD		2
 
-
 void CrosshairOn (void);
 void CrosshairOff (void);
 void HideCrosshair (void);
 void RestoreCrosshair (void);
+void DrawAttached (void);
+void DrawMark (void);
 void MoveCrosshairRelative (LocationType, LocationType);
 bool MoveCrosshairAbsolute (LocationType, LocationType);
 void SetCrosshairRange (LocationType, LocationType, LocationType,
 			LocationType);
 void InitCrosshair (void);
 void DestroyCrosshair (void);
-void DrawMark (void);
 void FitCrosshairIntoGrid (LocationType, LocationType);
 
 #endif
diff --git a/src/hid/lesstif/main.c b/src/hid/lesstif/main.c
index c5f3e3e..4595b80 100644
--- a/src/hid/lesstif/main.c
+++ b/src/hid/lesstif/main.c
@@ -2433,7 +2433,6 @@ idle_proc (XtPointer dummy)
       int mx, my;
       BoxType region;
       lesstif_use_mask (0);
-      Crosshair.On = 0;
       pixmap = main_pixmap;
       mx = view_width;
       my = view_height;
@@ -2519,7 +2518,8 @@ idle_proc (XtPointer dummy)
       XCopyArea (display, main_pixmap, window, my_gc, 0, 0, view_width,
 		 view_height, 0, 0);
       pixmap = window;
-      CrosshairOn ();
+      DrawAttached ();
+      DrawMark ();
       need_redraw = 0;
     }
